### ISSUE TITLE
How can I generate nonWitnessUtxo for signTransaction?

### ISSUE THREAD

**Issue by A2DNEW** [bitcoinjs/bitcoinjs-lib#2084]
I want to sign transaction locally and after that send it to network

1- i created key and address with function:

```
const keyPair = ECPair.makeRandom({ network: dogecoin });
const { address } = bitcoin.payments.p2pkh({ pubkey: keyPair.publicKey, network: dogecoin });
console.log('address', address); // DEnbFjV4kmEiUfrmY27seaLY9kW5r7YtZV
console.log('keyPair', keyPair.toWIF()); // QVTvEg3fkC5qYdkNRCB7TffWG3NBDzJvNEEpXhPendKL8Q3Y1FcN
```

2- i want to sign the transaction:

```
const alice = ECPair.fromWIF(
  'QVTvEg3fkC5qYdkNRCB7TffWG3NBDzJvNEEpXhPendKL8Q3Y1FcN',
  dogecoin,
);

const hash = '2a44f088778bee025f9c016880daac779be9f2e917f4363cb0dbe756514d6f18' // generate random
const data = '' // how can i generate it?!

const psbt = new bitcoin.Psbt({ network: dogecoin });
psbt.addInput({
  hash: hash,
  index: 0,
  nonWitnessUtxo: Buffer.from(
    data, 
    'hex',
  ),
});
```

how can i generate data for nonWitnessUtxo?

* please don't close my issue :-(

**Comment by junderw** [comment]
nonWitnessUtxo is the fully signed transaction with the ID of `2a44f088778bee025f9c016880daac779be9f2e917f4363cb0dbe756514d6f18`.

**Comment by A2DNEW** [comment]
> nonWitnessUtxo is the fully signed transaction with the ID of `2a44f088778bee025f9c016880daac779be9f2e917f4363cb0dbe756514d6f18`.

Thank you
But how?
Can you send me method or function for sign this ID? `2a44f088778bee025f9c016880daac779be9f2e917f4363cb0dbe756514d6f18`?

**Comment by A2DNEW** [comment]
@junderw can you help me?

**Comment by junderw** [comment]
You need coins to send coins.

If you receive coins, they are sent to you in an output inside a transaction.

If you want to spend those coins, you need to reference the transaction's output where you received the coins.

If you are XYZ and you receive a tx that says `TX ID ABC: "Send 2 coins to XYZ, then send the remainder to myself."`

When you want to send 1 coin, you need to make a tx that says "TX ID ???: "Send the 2 coins I got in TX ID ABC in the first output, and send 1 coin to ??? and send 1 coin back to myself"

So the txid needs to point to `ABC` in the above example. The index (aka vout) would be 0 since it's the first output.

tl;dr you need coins to send coins.

Dogecoin is basically a copy paste of Bitcoin, so please read this website to learn about how it works on a basic level.

https://developer.bitcoin.org/devguide/

**Comment by tandn37** [comment]
@A2DNEW you can get it from here
https://mempool.space/testnet/api/tx/608bf9fe61690c89064eefad62e14dea2974f7b7a7411b9f3b9efa5ee5732d9f/hex

**Comment by A2DNEW** [comment]
> @A2DNEW you can get it from here https://mempool.space/testnet/api/tx/608bf9fe61690c89064eefad62e14dea2974f7b7a7411b9f3b9efa5ee5732d9f/hex

thank you
how about dogecoin or litecoin?

**Comment by A2DNEW** [comment]
> You need coins to send coins.
> 
> If you receive coins, they are sent to you in an output inside a transaction.
> 
> If you want to spend those coins, you need to reference the transaction's output where you received the coins.
> 
> If you are XYZ and you receive a tx that says `TX ID ABC: "Send 2 coins to XYZ, then send the remainder to myself."`
> 
> When you want to send 1 coin, you need to make a tx that says "TX ID ???: "Send the 2 coins I got in TX ID ABC in the first output, and send 1 coin to ??? and send 1 coin back to myself"
> 
> So the txid needs to point to `ABC` in the above example. The index (aka vout) would be 0 since it's the first output.
> 
> tl;dr you need coins to send coins.
> 
> Dogecoin is basically a copy paste of Bitcoin, so please read this website to learn about how it works on a basic level.
> 
> https://developer.bitcoin.org/devguide/

Well, If I have 10,000 transactions, I have to get all of them and sign?

**Comment by jasonandjay** [comment]
> > You need coins to send coins.
> > If you receive coins, they are sent to you in an output inside a transaction.
> > If you want to spend those coins, you need to reference the transaction's output where you received the coins.
> > If you are XYZ and you receive a tx that says `TX ID ABC: "Send 2 coins to XYZ, then send the remainder to myself."`
> > When you want to send 1 coin, you need to make a tx that says "TX ID ???: "Send the 2 coins I got in TX ID ABC in the first output, and send 1 coin to ??? and send 1 coin back to myself"
> > So the txid needs to point to `ABC` in the above example. The index (aka vout) would be 0 since it's the first output.
> > tl;dr you need coins to send coins.
> > Dogecoin is basically a copy paste of Bitcoin, so please read this website to learn about how it works on a basic level.
> > https://developer.bitcoin.org/devguide/
> 
> Well, If I have 10,000 transactions, I have to get all of them and sign?

just got transactions can apply the amount you want to send

such as you have there transaction
1. output a with 10000sats amount
2. output b with 10000sats amount
3. output c with 5000sats amount

If you want send 2000, you just need output c
If you want send 8000, you just need one of a or b
If you want send 22000, you need a, b and c

**Comment by junderw** [comment]
> Well, If I have 10,000 transactions, I have to get all of them and sign?

If you want to send all your coins, then yes. That's how the UTXO model works.

If you only want to send 20 coins and one of those 10,000 transactions sent you 30 coins, you can choose to only add that one transaction's coins.

Then you create two outputs, one sending to the other person for 20, then another output sending to yourself with 10 minus transaction fee you want to add.

**Comment by A2DNEW** [comment]
> > > You need coins to send coins.
> > > If you receive coins, they are sent to you in an output inside a transaction.
> > > If you want to spend those coins, you need to reference the transaction's output where you received the coins.
> > > If you are XYZ and you receive a tx that says `TX ID ABC: "Send 2 coins to XYZ, then send the remainder to myself."`
> > > When you want to send 1 coin, you need to make a tx that says "TX ID ???: "Send the 2 coins I got in TX ID ABC in the first output, and send 1 coin to ??? and send 1 coin back to myself"
> > > So the txid needs to point to `ABC` in the above example. The index (aka vout) would be 0 since it's the first output.
> > > tl;dr you need coins to send coins.
> > > Dogecoin is basically a copy paste of Bitcoin, so please read this website to learn about how it works on a basic level.
> > > https://developer.bitcoin.org/devguide/
> > 
> > 
> > Well, If I have 10,000 transactions, I have to get all of them and sign?
> 
> just got transactions can apply the amount you want to send
> 
> such as you have there transaction
> 
> 1. output a with 10000sats amount
> 2. output b with 10000sats amount
> 3. output c with 5000sats amount
> 
> If you want send 2000, you just need output c If you want send 8000, you just need one of a or b If you want send 22000, you need a, b and c

> > Well, If I have 10,000 transactions, I have to get all of them and sign?
> 
> If you want to send all your coins, then yes. That's how the UTXO model works.
> 
> If you only want to send 20 coins and one of those 10,000 transactions sent you 30 coins, you can choose to only add that one transaction's coins.
> 
> Then you create two outputs, one sending to the other person for 20, then another output sending to yourself with 10 minus transaction fee you want to add.

Thank you
And my last question:

A send 20 coins to B and that txid is: XXX
If B wants to send 10 coins to C, He have to use XXX for sign and set 2 outputs: 10 coins for C and 10 coins for himself.
Thats right?
So, after that, for next transaction we can not use XXX again? and we have to use new txid (created with 10 coins)?

**Comment by jasonandjay** [comment]
This is roughly correct. I'll add two details.
1. UTXO represents unspent transaction output and cannot be reused. The next transaction has a new txid
2. You cannot give yourself 10 coins in change, you need to pay a transaction fee.
